File: src/main/java/de/nordgedanken/auto_hotkey/run_configurations/core/AhkRunConfig.java
package de.nordgedanken.auto_hotkey.run_configurations.core;

import com.intellij.execution.Executor;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.execution.configurations.RunConfigurationBase;
import com.intellij.execution.configurations.RunProfileState;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.roots.ui.configuration.IdeaProjectSettingsService;
import de.nordgedanken.auto_hotkey.AhkConstants;
import de.nordgedanken.auto_hotkey.run_configurations.execution.AhkRunState;
import de.nordgedanken.auto_hotkey.run_configurations.model.AhkRunConfigSettings;
import de.nordgedanken.auto_hotkey.run_configurations.ui.AhkRunConfigSettingsEditor;
import de.nordgedanken.auto_hotkey.sdk.AhkSdkType;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Defines instances of Ahk run configurations.
 */
@State(
	name = AhkConstants.PLUGIN_NAME,
	storages = {@Storage(AhkConstants.PLUGIN_NAME + "__run-configuration.xml")}
)
public class AhkRunConfig extends RunConfigurationBase<Object> {
	public AhkRunConfigSettings runConfigSettings = new AhkRunConfigSettings();

	protected AhkRunConfig(@NotNull Project project, @Nullable ConfigurationFactory factory, @Nullable String name) {
		super(project, factory, name);
	}

	@Override
	public @NotNull SettingsEditor<? extends RunConfiguration> getConfigurationEditor() {
		return new AhkRunConfigSettingsEditor(this.getProject());
	}

	@Override
	public @Nullable RunProfileState getState(@NotNull Executor executor, @NotNull ExecutionEnvironment environment) {
		Sdk projectSDK = ProjectRootManager.getInstance(getProject()).getProjectSdk();
		//validating the Ahk SDK here. Eventually this should be removed and validation should happen while editing the run config
		if(projectSDK == null || !(projectSDK.getSdkType() instanceof AhkSdkType)) {
			String fullMessage = "SDK Error: You must create an AutoHotkey SDK and select it as the project's default SDK in order to run the AHK script. <br>Otherwise IntelliJ does not know what to use to run your script";
			NotificationUtil.showErrorPopup("Execution Error", fullMessage, getProject(), environment);
			IdeaProjectSettingsService.getInstance(getProject()).openProjectSettings();
		} else if(runConfigSettings.getPathToScript() == null || runConfigSettings.getPathToScript().isEmpty()) {
			String fullMessage = "Error: You must specify the path to the script you want to execute within the run config";
			NotificationUtil.showErrorPopup("Execution Error", fullMessage, getProject(), environment);
		} else {
			//return execution runstate only if everything is good. Else return null so nothing happens
			return new AhkRunState(this, environment);
		}
		return null;
	}

	/**
	 * This READS any prior persisted configuration from the State/Storage defined by this classes annotations.
	 */
	@Override
	public void readExternal(@NotNull Element element) {
		super.readExternal(element);
		runConfigSettings.populateFromElement(element);
	}

	/**
	 * This WRITES/persists configurations TO the State/Storage defined by this classes annotations.
	 */
	@Override
	public void writeExternal(@NotNull Element element) {
		super.writeExternal(element);
		runConfigSettings.writeToElement(element);
	}

	@Override
	public String toString() {
		return "AhkRunConfig{" + "runConfigSettings=" + runConfigSettings + '}';
	}
}

File: src/main/java/de/nordgedanken/auto_hotkey/run_configurations/core/AhkRunConfig.kt
package de.nordgedanken.auto_hotkey.run_configurations.core

import com.intellij.execution.Executor
import com.intellij.execution.configurations.*
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.project.Project
import de.nordgedanken.auto_hotkey.AhkConstants
import de.nordgedanken.auto_hotkey.localization.AhkBundle
import de.nordgedanken.auto_hotkey.run_configurations.execution.AhkRunState
import de.nordgedanken.auto_hotkey.run_configurations.model.AhkRunConfigSettings
import de.nordgedanken.auto_hotkey.run_configurations.ui.AhkRunConfigSettingsEditor
import de.nordgedanken.auto_hotkey.sdk.getAhkSdkByName
import org.jdom.Element
import java.io.File

/**
 * Defines instances of Ahk run configurations.
 */
@State(name = AhkConstants.PLUGIN_NAME, storages = [Storage(AhkConstants.PLUGIN_NAME + "__run-configuration.xml")])
class AhkRunConfig(
        project: Project,
        factory: ConfigurationFactory?,
        name: String?
) : RunConfigurationBase<Any?>(project, factory, name) {
    val runConfigSettings = AhkRunConfigSettings()

    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration?> {
        return AhkRunConfigSettingsEditor(this.project)
    }

    override fun getState(executor: Executor, environment: ExecutionEnvironment): RunProfileState? {
        return AhkRunState(this, environment)
    }

    /**
     * Verifies that the run config settings are valid. (Prevents execution if not valid)
     */
    @Throws(RuntimeConfigurationException::class)
    override fun checkConfiguration() {
        if (getAhkSdkByName(runConfigSettings.runner) == null) {
            throw RuntimeConfigurationError(AhkBundle.msg("runconfig.configtab.error.scriptrunner.notahksdktype"))
        } else {
            val f = File(runConfigSettings.pathToScript)
            if (!f.exists()) {
                throw RuntimeConfigurationError(AhkBundle.msg("runconfig.configtab.error.scriptpath.notexist"))
            } else if (f.extension != AhkConstants.FILE_EXTENSION) {
                throw RuntimeConfigurationError(AhkBundle.msg("runconfig.configtab.error.scriptpath.notahkextension"))
            }
        }
    }

    /**
     * This READS any prior persisted configuration from the State/Storage defined by this classes annotations.
     */
    override fun readExternal(element: Element) {
        super.readExternal(element)
        runConfigSettings.populateFromElement(element)
    }

    /**
     * This WRITES/persists configurations TO the State/Storage defined by this classes annotations.
     */
    override fun writeExternal(element: Element) {
        super.writeExternal(element)
        runConfigSettings.writeToElement(element)
    }
}
